File: backend/src/main/kotlin/com/trackwatch/backend/clients/spotify/SpotifyUserApiClient.kt
package com.trackwatch.backend.clients.spotify

import com.trackwatch.backend.clients.spotify.dto.SpotifyUserDTO
import com.trackwatch.backend.exception.*
import com.trackwatch.backend.exception.ErrorCode
import com.trackwatch.backend.service.MetricService
import org.slf4j.LoggerFactory

import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.client.WebClientResponseException

@Component
class SpotifyUserApiClient(metricService: MetricService): SpotifyApiClient(metricService) {

    private val log = LoggerFactory.getLogger(SpotifyUserApiClient::class.java)

    fun getUser(accessToken: String): SpotifyUserDTO {
        try {
            sendMetricApiCall("getUser")

            val response = webClient.get()
                .uri("/me")
                .header("Authorization", "Bearer $accessToken")
                .retrieve()
                .bodyToMono(Map::class.java)
                .block() ?: throw UnauthorizedException(ErrorCode.SPOTIFY_INVALID_ACCESS_TOKEN)

            return mapToSpotifyUserDTO(response)

        } catch (e: WebClientResponseException) {
            throw handleSpotifyApiException(e)

        } catch (e: Exception) {
            throw InternalServerErrorException(ErrorCode.UNHANDLED_EXCEPTION, "Error while calling Spotify API", e.toString())
        }
    }

    private fun mapToSpotifyUserDTO(response: Map<*, *>): SpotifyUserDTO {
        val explicitContent = response["explicit_content"] as Map<*, *>
        val imagesJson = response["images"] as List<*>

        return SpotifyUserDTO(
            id = response["id"] as String,
            email = response["email"] as String,
            name = response["display_name"] as String,
            imageUrl = imagesJson.firstOrNull()?.let { (it as Map<*, *>)["url"] as String } ?: "",
            blockedExplicitContent = explicitContent["filter_enabled"] as Boolean
        )
    }

    private fun handleSpotifyApiException(e: WebClientResponseException): Exception {
        log.error(e.toString())

        return when (e.statusCode.value()) {
            401 -> UnauthorizedException(ErrorCode.SPOTIFY_INVALID_ACCESS_TOKEN, details = e.message)
            403 -> ForbiddenException(ErrorCode.SPOTIFY_FORBIDDEN_REQUEST, details = e.message)
            404 -> BadRequestException(ErrorCode.SPOTIFY_USER_NOT_FOUND)
            else -> InternalServerErrorException(ErrorCode.UNHANDLED_EXCEPTION, "Error while calling Spotify API: ${e.message}")
        }
    }

}

File: backend/app/clients/spotify/spotify_user_api_client.py
from .spotify_api_client import spotify_api_request
from app.exceptions import *

def get_spotify_user(access_token: str):
  try:
    response = spotify_api_request(
      method="GET",
      endpoint="/me",
      token=access_token
    )
    return map_to_spotify_user_dto(response)

  except Exception as e:
    if hasattr(e, "response") and hasattr(e.response, "status_code"):
      code = e.response.status_code
      msg = str(e)
      if code == 401: raise UnauthorizedException(ErrorCode.SPOTIFY_INVALID_ACCESS_TOKEN, details=msg)
      if code == 403: raise ForbiddenException(ErrorCode.SPOTIFY_FORBIDDEN_REQUEST, details=msg)
      if code == 404: raise BadRequestException(ErrorCode.SPOTIFY_USER_NOT_FOUND)

    raise InternalServerErrorException(
      ErrorCode.UNHANDLED_EXCEPTION,
      "Error while calling Spotify API",
      str(e)
    )

def map_to_spotify_user_dto(response: dict):
  images_json = response.get("images", [])
  user_id = response.get("id")
  email = response.get("email") or f"{user_id}@trackwatch.placeholder"

  return {
    "id": user_id,
    "email": email,
    "name": response.get("display_name"),
    "image_url": images_json[0]["url"] if images_json else "",
    "blocked_explicit_content": False
  }
